src/style-spec/validate/validate_expression.js

```diff
@@ -1,32 +1,21 @@
 
 const ValidationError = require('../error/validation_error');
-const {findZoomCurve, getExpectedType} = require('../expression');
-const compile = require('../expression/compile');
-const Curve = require('../expression/definitions/curve');
+const createExpression = require('../expression');
+const {getExpectedType, getDefaultValue} = require('../expression');
 const unbundle = require('../util/unbundle_jsonlint');
 
 module.exports = function validateExpression(options) {
-    const expression = deepUnbundle(options.value.expression);
-    const compiled = compile(expression, getExpectedType(options.valueSpec));
+    const expression = createExpression(
+        deepUnbundle(options.value.expression),
+        getExpectedType(options.valueSpec),
+        getDefaultValue(options.valueSpec));
 
-    const key = `${options.key}.expression`;
-
-    if (compiled.result === 'success') {
-        if (!options.disallowNestedZoom || compiled.isZoomConstant) {
-            return [];
-        }
-
-        const curve = findZoomCurve(compiled.expression);
-        if (curve instanceof Curve) {
-            return [];
-        } else if (curve) {
-            return [new ValidationError(`${key}${curve.key}`, options.value, curve.error)];
-        } else {
-            return [new ValidationError(`${key}`, options.value, '"zoom" expression may only be used as input to a top-level "curve" expression.')];
-        }
```

Comment:
I think this check makes sense here, rather than in `creatExpression`, because the latter will be used in contexts where the zoom curve restriction isn't applicable -- for example, expressions used for filters can use `[zoom]` anywhere.

Reply:
I think it's better to have validation and error generation centralized, so that the `"zoom"` check uses the same conventions and types as other static errors. To control where `"zoom"` is allowed we can use an option. I assume that's what the (currently unused) `disallowNestedZoom` option was -- when it's needed again, it can be passed on to `createExpression`.